zcncore/transaction.go

```diff
@@ -226,22 +228,47 @@ func (t *Transaction) completeVerifyWithConStatus(status int, conStatus Confirma
 	t.verifyConfirmationStatus = conStatus
 	t.verifyOut = out
 	t.verifyError = err
+	if status == StatusError {
+		transaction.Cache.Evict(t.txn.ClientID)
+	}
 	if t.txnCb != nil {
 		t.txnCb.OnVerifyComplete(t, t.verifyStatus)
 	}
 }
 
+type getNonceCallBack struct {
+	nonceCh chan int64
+	err     error
+}
+
+func (g getNonceCallBack) OnNonceAvailable(status int, nonce int64, info string) {
+	if status != StatusSuccess {
+		g.err = errors.New("get_nonce", "failed respond nonce")
+	}
+
+	g.nonceCh <- nonce
+}
+
 func (t *Transaction) submitTxn() {
-	// Clear the status, incase transaction object reused
+	// Clear the status, in case transaction object reused
 	t.txnStatus = StatusUnknown
 	t.txnOut = ""
 	t.txnError = nil
 
+	nonce := t.txn.TransactionNonce
+	if nonce < 1 {
+		nonce = transaction.Cache.GetNextNonce(t.txn.ClientID)
+	} else {
+		transaction.Cache.Set(t.txn.ClientID, nonce)
+	}
+	t.txn.TransactionNonce = nonce
+
```

Comment:
If the nonce was < 1, and we get value from cache, and update the transaction, shouldn't we sign this transaction again? maybe we should return error here instead of update the nonce. The caller should set the nonce and signed it correctly before submit.

Reply:
@peterlimg sorry, just found your comment. Do you mean that in case of transaction reuse we can get problems since nonce will be updated and signature won't?

Reply:
@dabasov I'm not sure if this is caused by reusing transaction, but the issue is we change the `transaction.Nonce` in `submitTxn()` method but the signature is not updated. We'd not changing any fields when submitting a transaction, it may cause unexpected result.

Reply:
It will be signed if signature is not provided and it is not provided normally, I guess the only way transaction is signed that it is reused. I think we can force resign here